src/storage/src/hummock/sstable/writer.rs

```diff
@@ -0,0 +1,166 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use bytes::{BufMut, Bytes, BytesMut};
+use risingwave_hummock_sdk::HummockSstableId;
+
+use crate::hummock::{
+    CachePolicy, HummockResult, SstableBuilderOptions, SstableStoreRef, SstableStreamingUploader,
+};
+
+/// A consumer of SST data.
+/// The data may be written to a stream, pushed to a vector, appended to a buffer, etc.
+#[async_trait::async_trait]
+pub trait SstableWriter: Send {
+    type Output;
+
+    /// Write an SST block to the writer.
+    fn write_block(&mut self, block: &[u8]) -> HummockResult<()>;
+
+    /// Get the length of data that has already been written.
+    fn data_len(&self) -> usize;
+
+    /// Finish writing the SST. Return the output along with final data length.
+    async fn finish(mut self, size_footer: u32) -> HummockResult<(usize, Self::Output)>;
+}
+
+pub struct InMemSstableWriter {
+    buf: BytesMut,
+}
+
+/// Append SST data to a buffer.
+impl InMemSstableWriter {
+    pub fn new(capacity: usize) -> Self {
+        Self {
+            buf: BytesMut::with_capacity(capacity),
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl SstableWriter for InMemSstableWriter {
+    type Output = Bytes;
+
+    fn write_block(&mut self, block: &[u8]) -> HummockResult<()> {
+        self.buf.put_slice(block);
+        Ok(())
+    }
+
+    async fn finish(mut self, size_footer: u32) -> HummockResult<(usize, Self::Output)> {
+        self.buf.put_slice(&size_footer.to_le_bytes());
+        Ok((self.buf.len(), self.buf.freeze()))
+    }
+
+    fn data_len(&self) -> usize {
+        self.buf.len()
+    }
+}
+
+impl From<&SstableBuilderOptions> for InMemSstableWriter {
+    fn from(options: &SstableBuilderOptions) -> InMemSstableWriter {
+        InMemSstableWriter::new(options.capacity + options.block_capacity)
+    }
+}
+
+/// Upload sst blocks to the streaming uploader of `SstableStore`.
+/// The upload might not be finished immediately on calling `finish`.
+/// Instead, it returns a sealed uploader which is ready to be passed back to `SstableStore` to
+/// finish the uploading.
+pub struct StreamingSstableWriter {
+    uploader: SstableStreamingUploader,
+    data_len: usize,
+}
+
+impl StreamingSstableWriter {
+    pub fn new(uploader: SstableStreamingUploader) -> StreamingSstableWriter {
+        Self {
+            uploader,
+            data_len: 0,
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl SstableWriter for StreamingSstableWriter {
+    type Output = SstableStreamingUploader;
+
+    fn write_block(&mut self, block: &[u8]) -> HummockResult<()> {
+        self.data_len += block.len();
+        self.uploader.upload_block(Bytes::from(block.to_vec()))
```

Comment:
We may turn a `&[u8]` into `Bytes` directly without turning it into `Vec` first.

Reply:
Copying `block` is necessary, because `BlockBuilder` will clear its own buffer and write new data into it after the current block is written. `Bytes::from(block.to_vec())` copies the data only once, which I think is already optimal.

Reply:
Bytes may already provide method to be initialized from a `&[u8]` directly without turning it into a `Vec`, though this method also involves a memory copy.

Reply:
I didn't find such a method on `Bytes` 😢

Reply:
I see that `Bytes` implements `From<&'a [u8]>`. Maybe we can take a try.

https://docs.rs/bytes/0.4.12/bytes/struct.Bytes.html#impl-From%3C%26%27a%20%5Bu8%5D%3E

Reply:
This method is no longer available on `1.0.0` version of `bytes` crate. But I've found a ~15% faster alternative: `BytesMut::from(block).freeze()`